pkg/controller/controller.go

```diff
@@ -1175,20 +1183,26 @@ func (p *csiProvisioner) Delete(ctx context.Context, volume *v1.PersistentVolume
 }
 
 func (p *csiProvisioner) canDeleteVolume(volume *v1.PersistentVolume) error {
-	if p.vaLister == nil {
-		// Nothing to check.
-		return nil
-	}
+	if p.vaLister != nil {
+		// Verify if volume is attached to a node before proceeding with deletion
+		vaList, err := p.vaLister.List(labels.Everything())
+		if err != nil {
+			return fmt.Errorf("failed to list volumeattachments: %v", err)
+		}
 
-	// Verify if volume is attached to a node before proceeding with deletion
-	vaList, err := p.vaLister.List(labels.Everything())
-	if err != nil {
-		return fmt.Errorf("failed to list volumeattachments: %v", err)
-	}
+		for _, va := range vaList {
+			if va.Spec.Source.PersistentVolumeName != nil && *va.Spec.Source.PersistentVolumeName == volume.Name {
+				return fmt.Errorf("persistentvolume %s is still attached to node %s", volume.Name, va.Spec.NodeName)
+			}
+		}
+	} else if p.volumesInUseProtection {
```

Comment:
We only chceck if vaLister == nil?
if vaLister != nil and no VA matches volume name, we also need to check volumeInUse?

Reply:
> if vaLister != nil and no VA matches volume name, we also need to check volumeInUse?

+1, for driver that is non-attachable, do we create VA? If not, then we need to check even if there is no VA, is the volume still in use

Reply:
I think we should make it more explicit that no vaLister <=> no controller publish/unpublish. Right now it only seems documented at csi_provisioner.go:270 and it would be easy to break that assumption no knowing the implication here.

So maybe adding a flag in csiProvisioner would make that more clear?

Reply:
For the case when vaLister !=nil, ie we expect to have a VA, in https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_attacher.go#L359 it seems pretty clear if the VA is deleted then the volume is not in use... at least the csi controller detach has finished successfully, isn't that enough?

Reply:
@mattcary  yes if vaLister != nil (which implies volume supports VA creation), and the absence of a VA is a definite signal that controller unpublish was successful (and that indirectly implies unpublish/unstage succeeded)

Reply:
From other comments it's clear that vaLister indicating if the driver is attachable is a little subtle. Can you document that a bit more in the code to avoid future confusion?

Reply:
Ack